refactor: Refactor compose files to using multi-file approach

lenny-intel opened the pull request:
This removes majority of the duplication, leaving only some duplication of env overrides.

Please pull this down and give it a run. See README for run options

Please try it out on your systems. Related PR for BB Test issue (https://github.com/edgexfoundry/blackbox-testing/issues/452) has to be ready before this can be merged.

close #281

bnevis-i commented on releases/nightly-build/compose-files/README.md:
```diff
@@ -0,0 +1,44 @@
+
+This folder contain the docker compose files that pull and run the EdgeX images from the Nexus3 docker registry that are tagged `master`. These images are build from the Edgex CI Pipeline when PRs are merged into the `master` branch.
+
+> *Note: Docker does not re-pull newer instances of these images. You must explicitly remove old instances so the new instances are pulled. See the `make remove` command below that will do this for you.*
```
Why not just have a rule that does the docker-compose pull?

lenny-intel replied:
I considered that, but `pull` which config? Pull for all configs, isn't efficient? Didn't want a buch of `pull` variations. This makes sure when you run your config the images it uses are freshly pulled.

bnevis-i replied:
Ah, I see.  Perhaps a reasonable middle ground would be to pull all containers per arch, whether or not they are actually run.  Or maybe a makefile isn't the right option here.  Maybe it should just be a shell script that simply computes the .yml files needed based on command line options, computes the architecture by scrutinizing the output of uname -a, and then passes additional flags such as up, down, pull, etc to docker-compose?

bnevis-i commented on releases/nightly-build/compose-files/Makefile:
```diff
@@ -0,0 +1,91 @@
+.PHONY: portainer portainer-down run run-no-sec run-mqtt run-mqtt-no-sec run-arm run-arm-no-sec run-mqtt-arm run-mqtt-arm-no-sec down clean remove
+
+portainer:
```
It would be helpful to have a "help:" target at the front that prints a brief usage instructions (such listing the available targets).  As it is, "make" would simply start portainer as the default target.

lenny-intel replied:
I like it!

lenny-intel replied:
Added `help` target.

bnevis-i commented on releases/nightly-build/compose-files/Makefile:
```diff
@@ -0,0 +1,91 @@
+		down
+
+clean: down
+	-docker rm $$(docker ps -aq)
```
This line is hazardous to my system as I have other containers running beside edgex's.  It is trying to remove these other containers.

lenny-intel replied:
This only removes them if the container is stopped. I can look at refining it, but need to ensure that all the containers used be the stack are removed before pruning. otherwise the prune will not get everything.

bnevis-i replied:
I have observed that there is no ill side effects to running docker-compose "down" on stacks that I haven't done an "up" on.  In other words, why not "docker-compose -f *.yml down"

lenny-intel replied:
I can filer on the network used to just hit the edgex stack containers.

bnevis-i replied:
While you're at it, this might be a good time to explicitly name the network instead of inheriting it from cwd.

lenny-intel replied:
Already done. By specifying the `-p edgex` on the docker-compose command. This puts all the containers, networks, volumes, etc. under the `edgex` stack and names them with the `edgex_` prefix.

lenny-intel replied:
> I have observed that there are no ill side effects to running docker-compose "down" on stacks that I haven't done an "up" on. In other words, why not "docker-compose -f *.yml down"

Nice idea, but doesn't work as the `-f` needs to be repeated with each file. Some complex scripting could create command properly, but IMO this work fine with just 4 files. More files might justify a more dynamic approach.

jpwhitemn commented on releases/nightly-build/compose-files/README.md:
```diff
@@ -0,0 +1,44 @@
+
+These approach used for these compose files is the `Extending using multiple Compose files` described here: https://docs.docker.com/compose/extends/#multiple-compose-files
+
+The `Extending using multiple Compose files` approach removes the majority of the duplication in the multiple composes file that previous Edgex release compose file endure. This approach does make actual running of the solution more complicated due to having to list the multiple compose file required to run a particular configuration. To help alleviate this complexity, a `Makefile`has been provid ...
```
approach removes the majority of the duplication *found in previous EdgeX compose files.*
This approach *makes running the solution* more complicated ...

lenny-intel replied:
fixed

jpwhitemn commented on releases/nightly-build/compose-files/README.md:
```diff
@@ -0,0 +1,44 @@
+| make run-ui-arm          | Runs the optional **ARM64** EdgeX UI service.                |
+| make down                | Stops all EdgeX no matter which configuration they are running in. |
+| make clean               | Runs **`down`** and then removes any stopped containers, prunes unused volumes and networks |
+| make remove              | Runs **`clean`** and then removes all docker images with the **`master`** tag. This is how you make sure you are using the very latest **`master`** images. It is highly recommended to run this command prior to any validation of you current pull request. |
```
It is highly recommended *you* run this command prior to any validation of *your* current pull request.

lenny-intel replied:
fixed

lenny-intel commented on releases/nightly-build/compose-files/Makefile:
```diff
@@ -0,0 +1,91 @@
+		-f docker-compose-nexus-ui.yml \
+		down
+
+clean: down
+	-docker rm $$(docker ps -aq)
```
the leading `-` suppresses the error and I have add piping STDERR to null so thos messages are now seen. ` 2> /dev/null` now only left with `make: [clean] Error 1 (ignored)` which is ok, IMO

jim-wang-intel replied:
ok

lenny-intel commented on releases/nightly-build/compose-files/Makefile:
```diff
@@ -0,0 +1,91 @@
+	docker volume prune -f && \
+	docker network prune -f
+
+#TODO: When cutting next release, remove this make target as it only needed when release = "master".
+remove: clean
+	docker rmi -f $$(docker images | grep master | awk '{print $$3}')
```
I added `-` to suppress error and `2> /dev/null` to hide messages.

lenny-intel commented on releases/nightly-build/compose-files/README.md:
```diff
@@ -0,0 +1,44 @@
+
+> *Note: Docker does not re-pull newer instances of these images. You must explicitly remove old instances so the new instances are pulled. See the `make remove` command below that will do this for you.*
+
+These approach used for these compose files is the `Extending using multiple Compose files` described here: https://docs.docker.com/compose/extends/#multiple-compose-files
```
fixed

jpwhitemn commented:
@lenny-intel - ran through this morning.  Couldn't test ARM and didn't test MQTT.  Otherwise seemed to work great.  Here are some additional comments based on actually using he make:
- All the WARNINGS coming of each run command because ARCH, REDIS_ARCH, KONG_UBUNTU are not set and default to a blank string was a bit jarring.  Wonder if we should run with compose commands with --log-level=ERROR.  Maybe even have a Makefile parameter that defaults to ERROR but can be overriden?
- Should we have make pull-XXX options for all the configurations.  I find I typically want to pull before I do an up -d (just to make sure I have everything pulled ok first)
- We have a portainer down but not a UI down?  I guess that is because we include UI in the regular down.  Seems these should be treated consistently one way or the other.
- on remove, can't we just do a docker system prune --all --volumes

lenny-intel commented:
> * All the WARNINGS coming of each run command because ARCH, REDIS_ARCH, KONG_UBUNTU are not set and default to a blank string was a bit jarring.  Wonder if we should run with compose commands with --log-level=ERROR.  Maybe even have a Makefile parameter that defaults to ERROR but can be overriden?

I just pushed an update that sets these variables to blank for the non-arm commands so the warning is avoided.

> * Should we have make pull-XXX options for all the configurations.  I find I typically want to pull before I do an up -d (just to make sure I have everything pulled ok first)

The `remove` command is more efficient at achieving the same result as it only targets the images tagged with `master` that you have loaded and causes them to be reloaded, if needed on the next run.

> * We have a portainer down but not a UI down?  I guess that is because we include UI in the regular down.  Seems these should be treated consistently one way or the other.

Correct, the UI is included in the `down` command, so no need for `ui-down` command

> * on remove, can't we just do a docker system prune --all --volumes

`docker system prune` is a big hammer. My current `remove` is surgical as it only removes the images tagged with `master` as those are the ones that the CI pipeline is changing.

bnevis-i commented:
As I was looking into something today, I noticed that I have a use case that the current script currently doesn't cover.  I often start the framework, but not the WHOLE framework.  For example, if I am testing something for Kong, I will do a docker-compose up -d "kong" and stop there.  Could the makefile take a TARGET parameter or something like that, and only go up to that target.

lenny-intel commented:
> Could the makefile take a TARGET parameter or something like that, and only go up to that target.

@bnevis-i , I'll look into that next week.

bnevis-i commented:
@cloudxxx8 You have been silent.  What do you need for testing?

lenny-intel commented:
> > Could the makefile take a TARGET parameter or something like that, and only go up to that target.
> 

@bnevis-i , I figured it out. Now all the extras  on the run targets are interpreted as `services` to run. Give it a try.

lenny-intel commented:
@bnevis-i , @jim-wang-intel , @jpwhitemn , I have reworked the makefile to reduce the `run' targets down to one with `options`. Please take a look.

hutchic commented on releases/nightly-build/compose-files/docker-compose-nexus-add-device-services.yml:
```diff
@@ -0,0 +1,122 @@
+
+services:
+  device-virtual:
+    image: ${REPOSITORY}/docker-device-virtual-go${ARCH}:master
```
placing all versions ie `:master` into a `.env` will make releases easier. Update the `.env` to pin the versions and then be done

additionally allows folks (and CI) to `export CONSUL_VERSION=1.2.3` and check everything against specific version of thing X

rsdmike replied:
I think we need to leverage .env files a lot more than we are today for sure.

lenny-intel replied:
I played with this and LIKE  IT! Thanks @hutchic.

lenny-intel replied:
Done

hutchic commented on releases/nightly-build/compose-files/docker-compose-nexus-add-device-services.yml:
```diff
@@ -0,0 +1,122 @@
+#      SERVICE_HOST: edgex-device-snmp
+#    depends_on:
+#      - data
+#      - command
```
anyway for this to be handled more elegantly?

lenny-intel replied:
Handle what more elegantly?

lenny-intel replied:
The commented out services... ;-)

hutchic replied:
yeah exactly that. Rubber ducking: could these by moved into their own compose files? Not a blocker just talking out loud

lenny-intel commented on releases/nightly-build/compose-files/Makefile:
```diff
@@ -0,0 +1,129 @@
+endif
+ifeq (arm64, $(filter arm64,$(ARGS)))
+	ARCH:=-arm64
+	REDIS_ARCH:=arm64v8/
```
@hutchic , are you saying that the `redis:5.0.8-alpine` image works for x86 as well as ARM?

hutchic replied:
Yup exactly 👍 

![image](https://user-images.githubusercontent.com/697188/85031193-63e7fc00-b14c-11ea-853b-d226fa2a7229.png)

Docker will pull / run the correct image that matches your architecture one shouldn't need to pin the architecture

lenny-intel replied:
Interesting. @andresrinivasan , is there another reason you originally choose `arm64v8/redis:5.0.8-alpine` for the ARM compose files rather than the `redis:5.0.8-alpine`?

andresrinivasan replied:
Good catch. Previously, the standard image did not include ARM64. Very happy to see we're including that architecture now.

lenny-intel replied:
Ok then... switching to just the one image. THX @hutchic for catching this!

lenny-intel commented on releases/nightly-build/compose-files/Makefile:
```diff
@@ -0,0 +1,129 @@
+portainer-down:
+	docker-compose -p portainer -f docker-compose-portainer.yml down
+
+run:
```
This causes the images to be pulled every time I run, which I don't want as I bring the stack up and down many times during a testing and debug session. Also, not something needed when these files are pushed to a release folder since the images don't changes like they are for our `master` tags.

lenny-intel replied:
I did make the change to use `gen` and then simply run the generated compose file. See latest changes.

lenny-intel commented:
> ease of use change suggestion especially for CI would be have `run` do `gen` and then `docker-compose` without `-f`

Interesting suggestion... We have scripts in BB Tests and TAF that use these files. TAF was the reason I created `gen`. The script uses `gen`, not `run` and then move the compose file to the name and location the remaining scripts already expect. BB Tests deploy script simply call `make run`. So this change would simplify my makefile and not change how the scripts use it. I am intrigued... ;-) I will play around with this idea.

lenny-intel commented:
@cloudxxx8 , can you please verify once more. Lots of little changes. THX!

jim-wang-intel commented on releases/nightly-build/compose-files/docker-compose-nexus-base.yml:
```diff
@@ -0,0 +1,218 @@
+    networks:
+      - edgex-network
+    environment:
+      # KUIPER_DEBUG: "true"
```
commented out means the default value is `true`? or the default value is false?

jim-wang-intel commented on releases/nightly-build/compose-files/docker-compose-nexus-base.yml:
```diff
@@ -0,0 +1,218 @@
+    networks:
+      - edgex-network
+    environment:
+      # KUIPER_DEBUG: "true"
```
commented out means the default value is `true`? or the default value is false?

lenny-intel replied:
Default is `false`. If you want DEBUG logs then you un-comment this line.

lenny-intel commented:
@rsdmike , can we iterate on this with follow on issues and PRs to address IDE integration prior to release?

hutchic commented:
@rsdmike you can still use in your IDE do a `make gen` first and then all the typical `docker-compose` files are at your disposal 👍

rsdmike commented:
Make isn't really all that friendly or familiar for a windows environment. One of the core architectural tenets of this project. running `make gen` gives me: 
```
PS F:\Development\developer-scripts\releases\nightly-build\compose-files> make gen
DEV= \
CORE_EDGEX_REPOSITORY=nexus3.edgexfoundry.org:10004 \
ARCH= \
KONG_UBUNTU= \
docker-compose -p edgex -f docker-compose-nexus-base.yml -f docker-compose-nexus-add-device-services.yml -f docker-compose-nexus-add-security.yml config > docker-compose.yml
'DEV' is not recognized as an internal or external command,
operable program or batch file.
make: *** [Makefile:73: gen] Error 1
```

lenny-intel commented:
@rsdmike , works fine for me from my windows bash and powershell started from my bash, but I see same issue from windows terminal and powershell started from that terminal. Lets work on these `Windows` issue in next round of PRs.

rsdmike commented:
I want to be clear that I see this as being useful and none of my suggestions suggest that this _shouldnt_ be done at all. I think I see a few different ways that could easily be apart of this PR to help address this issue:
1) Move everything in nightly to the root of the repo. It doesn't make sense to have the nightly build in the releases folder along side the other _actual_ releases anyway. But this would make the issue go away which stems from that the docker extensions assumes your .envs are at the root of the repo. 
2) Check in the compose(s) generated by "make-gen" / provide a default docker-compose for the nightly. Basically keeping both the old way and the new way. (places these into a folder - could even add these to a CI job).
3) Move the .env files to the root of the repo. Shouldn't be an issue since nothing else is using them. Though this  moves us in the direction of tagging/releasing this repo.
4) Check in editor specific configs that have commands already configured allowing to customize commands to specify where the .env files are.
5) Create a different repo called "user-scripts" which isn't intended for developers and is meant for users to run EdgeX without worrying about having developer tools installed.
6) Don't do this here- there is now a 2-3 page readme on how to run docker-compose. The opposite of easy-to-use. I'm struggling to find other open source projects that have done this as well. Additionally -- this really only results in a reduction of two files since we'll no longer be releasing mongo. My bigger issue this,  this is now an artifact generation system being shipped in a repo that we currently point our users to. If this repo is meant for releases as it is now, then this CI/CD should gen the various compose files and make those available. At the very least, it shouldn't be in the releases folder.

lenny-intel commented:
> I want to be clear that I see this as being useful and none of my suggestions suggest that this _shouldnt_ be done at all. I think I see a few different ways that could easily be apart of this PR to help address this issue:
> 
> 1. Move everything in nightly to the root of the repo. It doesn't make sense to have the nightly build in the releases folder along side the other _actual_ releases anyway. But this would make the issue go away which stems from that the docker extensions assumes your .envs are at the root of the repo.

With doing #2 below the .env issue goes away as the generated compose files have all the variables resolved.

> 2. Check in the compose(s) generated by "make-gen" / provide a default docker-compose for the nightly. Basically keeping both the old way and the new way. (places these into a folder - could even add these to a CI job).

I have been think about this and agree. These generated and commit files are there for convenience and should never to be edited directly, but rather regenerated when the other files are edited.

> 3. Move the .env files to the root of the repo. Shouldn't be an issue since nothing else is using them. Though this  moves us in the direction of tagging/releasing this repo.

Not needed with #2 

> 4. Check in editor specific configs that have commands already configured allowing to customize commands to specify where the .env files are.

Not needed with #2 

> 5. Create a different repo called "user-scripts" which isn't intended for developers and is meant for users to run EdgeX without worrying about having developer tools installed.

Agreed, but beyond scope of this PR.

> 6. Don't do this here- there is now a 2-3 page readme on how to run docker-compose. The opposite of easy-to-use. I'm struggling to find other open source projects that have done this as well. Additionally -- this really only results in a reduction of two files since we'll no longer be releasing mongo. My bigger issue this,  this is now an artifact generation system being shipped in a repo that we currently point our users to. If this repo is meant for releases as it is now, then this CI/CD should gen the various compose files and make those available. At the very least, it shouldn't be in the releases folder.

This PR is not about reducing number of files. It is about removing the duplication across the files. Now when we have a change it only occurs in one file rather then 8.

lenny-intel commented:
Put this PR back to Draft until I am back from vacation and we can discuss @rsdmike concerns

lenny-intel commented on releases/nightly-build/compose-files/.env:
```diff
@@ -0,0 +1,41 @@
+CORE_EDGEX_VERSION=master
+CONSUL_VERSION=master
+VAULT_VERSION=1.3.1
+KUIPER_VERSION=0.3.2
```
Good catch!! Removed the first one.

lenny-intel commented on releases/nightly-build/compose-files/Makefile:
```diff
@@ -0,0 +1,112 @@
+	docker-compose -p edgex down
+
+clean: down
+	-docker rm $$(docker ps --filter "network=edgex_edgex-network" --filter "network=edgex_default" -aq) 2> /dev/null
```
Yes, but that is how docker builds the networks name when we use `-p edgex` for the stack name. Alternative is to change the network name in the compose file, but then it isn't explicit when someone runs the new committed generated single file compose files w/o  `-p edgex`

lenny-intel replied:
Leaving as is for now.

lenny-intel commented on releases/nightly-build/compose-files/Makefile:
```diff
@@ -0,0 +1,112 @@
+ifneq (no-ds, $(filter no-ds,$(ARGS)))
+	COMPOSE_FILES:=${COMPOSE_FILES} -f docker-compose-nexus-add-device-services.yml
+endif
+ifeq (no-secty, $(filter no-secty,$(ARGS)))
```
Added comment

lenny-intel commented on releases/nightly-build/compose-files/Makefile:
```diff
@@ -0,0 +1,112 @@
+endif
+ifeq (mqtt, $(filter mqtt,$(ARGS)))
+	ifeq (YES,$(SECURITY))
+		COMPOSE_FILES:=${COMPOSE_FILES} -f docker-compose-nexus-add-mqtt.yml -f docker-compose-nexus-add-security.yml
```
Good catch! Reordered so security is added last so check when adding MQTT isn't needed.

lenny-intel commented on releases/nightly-build/compose-files/common-security.env:
```diff
@@ -0,0 +1,25 @@
+#
+
+EDGEX_SECURITY_SECRET_STORE=true
+SECRETSTORE_HOST=edgex-vault
```
Nope, it is very Vault specific, but these are where you would `configure` the specifics.

lenny-intel commented on releases/nightly-build/compose-files/docker-compose-nexus-add-security.yml:
```diff
@@ -0,0 +1,286 @@
+version: '3.7'
+
+# REDIS5_PASSWORD_PATHNAME must have the same value as
+# security-secretstore-read/res/configuration.toml SecretStore.Passwordfile. Note edgex-go issue
```
This has been removed/replaced by database.env

lenny-intel commented on releases/nightly-build/compose-files/README.md:
```diff
@@ -0,0 +1,113 @@
+Options:
+	dev:    Generates a Kong access token using local dev built docker image
+			'make docker', which creates docker images tagged with 'master-dev'
+	arm64:  Generates a Kong access token using ARM64 image
```
Yes, thus the `arm64` option. The option work together or on own.

lenny-intel commented:
> that I never saw an instance of "extends" as in extending

@mkbhanda , thanks for the review. I have already refactored so that the single all in one compose files for the 4 different variations we support are generated and committed. As for use of `extends`..  It is designed for a different purpose than multi file. It allows you to declare an additional service that extends an already defined service, which is not what we are doing. What we are doing with the multi file approach is add new services and overriding settings for previously defined services. 

I have made updates per most of you comments. Please take a look at the latest. Note that I still need to rework the README a bit to adjust for the generation of the single file compose files and probably add a new README that sites with those compose files and the Makefile that generated them. S

lenny-intel commented:
> What do you think of moving `clean` and `down` in the nightly-build/compose-files/Makefile instead of being part of the file in `source` directory ?

I was playing around with the up and `down` from that Makefile, hand not considered the `clean`. `up` is hard as don't know which one... but down is possible as just use to secure as it is a super set... THX!

lenny-intel commented:
@difince , I have reworked Makefile to be only about run/pull/stop/clean using generated files. Also added README. The Makefile in source folder is for build/run/pull/stop/clean using the source files. Please check it out once more.

difince commented:
Previously, I suggested moving `clean` and `down` in the upper lever makefile ..but i have missed that they are going to work only if the containers are brought up via the makefile run.  

So now I would suggest to keep it simple - lets, let the developers bring up the compose file they want by themselves ..and clean them appropriately. In my opinion Makefile should provide means for generating appropriate compose file and getting token.

difince commented on releases/nightly-build/compose-files/source/Makefile:
```diff
@@ -0,0 +1,128 @@
+		--volume /tmp/edgex/secrets/edgex-security-proxy-setup:/tmp/edgex/secrets/edgex-security-proxy-setup:ro,z \
+		$(CORE_EDGEX_REPOSITORY)/docker-edgex-security-proxy-setup-go$(ARCH):$(CORE_EDGEX_VERSION)$(DEV) \
+		--init=false --useradd=developer --group=admin \
+		| grep " access token for developer" | sed 's/.*: \([^.]*\.[^.]*\.[^.]*\).*/\1/'
```
grep " access token for developer"  -> should be "access token for user developer"

lenny-intel replied:
fixed

lenny-intel commented on releases/nightly-build/compose-files/Makefile:
```diff
@@ -0,0 +1,71 @@
+	docker-compose -p portainer -f docker-compose-portainer.yml down
+
+pull:
+	docker-compose -f docker-compose-nexus${SECURITY}${ARM64}.yml pull ${SERVICES}
```
fixed

lenny-intel commented on releases/nightly-build/compose-files/Makefile:
```diff
@@ -0,0 +1,71 @@
+	docker-compose -f docker-compose-nexus${SECURITY}${ARM64}.yml pull ${SERVICES}
+
+run:
+	docker-compose -p edgex -f docker-compose-nexus${SECURITY}${ARM64}.yml up -d ${SERVICES}
```
fixed

lenny-intel commented:
> So now I would suggest to keep it simple - lets, let the developers bring up the compose file they want by themselves ..and clean them appropriately. In my opinion Makefile should provide means for generating appropriate compose file and getting token.

@difince,  Developers/users can still bring up the compose files anyway they want or choose to use the make commands for ease. I moved the build commands to the source/Makefile since that is where a contributors will be working on the source files.  I do see the possible need to also run it from the other Makefile so I have added `build` to it that simply calls `make build` on the source/Makefile.

difince commented on releases/nightly-build/compose-files/source/Makefile:
```diff
@@ -0,0 +1,128 @@
+	MQTT:=-mqtt
+endif
+ifeq (no-secty, $(filter no-secty,$(ARGS)))
+	SECURITY:="NO"
```
SECURITY:="NO" is not used and could be removed

lenny-intel replied:
Cleaned up. THX!

difince commented on releases/nightly-build/compose-files/source/Makefile:
```diff
@@ -0,0 +1,128 @@
+	SECURITY:="NO"
+	NO_SECURITY:=-no-secty
+else
+	SECURITY:="YES"
```
SECURITY:="YES" is not used and could be removed

lenny-intel replied:
Cleaned up. THX!

difince commented:
I just do not see it easier to use the makefile for bringing up the services especially if I want to bring up just some of them. Services names should be sent as an argument,  but I do not know their  names by heart and I miss the auto complete.    
But @lenny-intel, you are right. It is always better everyone to have possibility to choose their own way of work.

hutchic commented:
I know it wasn't the intention but line count sure did climb since my previous approval 😆 

IMO I'd still leave the versions in an `.env`. Makes releases trivial, makes overriding specific images easy `export` vs `vim`

lenny-intel commented:
> IMO I'd still leave the versions in an `.env`.

@hutchic , the versions are in the .env file within the source folder so creating next release is copy files, update .env and generate single file compose files. We'll never be editing these generated files.
